Implement ProjectsService and move Core services (Teams, Favourites) to a separate DevOpsClient type

pbelousov opened the pull request:
Hi @benmatselby,

This pull request describes what I meant by the issue #30.

I implemented a new 'DevOpsClient' type which contains services which relate to entire DevOps account and not to the project - Favourites, Projects, Teams. These APIs described in the 'Core' section of the documentation: https://docs.microsoft.com/ru-ru/rest/api/azure/devops/core/

I also renamed 'Client' type to 'ProjectClient' as all its services are about just one project.

I didn't remove NewClient(account, project, token) function for backward compatibility, but it also could be initialized something like this:

```go
client := azuredevops.NewDevOpsClient(account, token)
projectClient := client.NewProjectClient(project)
```

Looking forward to your feedback!

Thank you!

benmatselby commented:
Hey @pbelousov Nice work, lot's of effort in there. And this totally makes sense, and an oversight from me. Thanks for raising.

Do you think one client is better? The main reason for this library was for me to build http://github.com/benmatselby/donny and when I pull in this change, it means I need to refactor things, which isn't an issue. However, it means I would kind of need two clients within my app. I wonder if we could refactor into one client?

In the mean time i've created a release of v0.1.0 so folks can keep with the old structure, if we want to refactor this into one client and accept a breaking change.

How you feel about that?

pbelousov commented:
Thank you for the answer and your feedback!

I already implemented something similar in C# a year ago. I've reviewed the code and I had just one client. But it was just a simple console app with 4-5 methods of Azure DevOps API and it was OK to have just one client.
 
But in the case of one client in a generic library, it could be a bit strange to pass a project name or project ID to all of the methods related to the project.
For example, an option with one client:
```go
client := azuredevops.NewClient(account, token)
projects := client.Projects.List()
repos := client.Git.List(projects[0].Name)
buildDefinitions := client.BuildDefinition.List(project[0].Name)
```
It looks like most of current methods will start using 'projectName' parameter and it could look a bit strange.

So maybe the approach with two clients is not so bad?

Or maybe 'Project' type could have all these services? Something like this:
```go
client := azuredevops.NewClient(account, token)
projects := client.Projects.List()
project := project[0]
repos := project.Git.List()
buildDefinitions := project.BuildDefinition.List()
```

What do you think about this?

benmatselby commented:
Hey @pbelousov Thanks for the input.. Let me have a think.. I'm currently not using Azure DevOps at the moment, so I need to get all my credentials back in order to test this 😄

benmatselby commented on azuredevops/azuredevops.go:
```diff
@@ -17,69 +17,98 @@ const (
-func NewClient(account string, project string, token string) *Client {
-	c := &Client{
+// NewClient gets a new Project Client for Azure DevOps API
+func NewClient(account string, project string, token string) *ProjectClient {
```
Can this be renamed to `NewProjectClient` please? There are breaking changes throughout this PR, so to keep `NewClient` as is, doesn't really add value. I'd prefer to be explicit please.

benmatselby commented on azuredevops/azuredevops_test.go:
```diff
@@ -71,17 +79,21 @@ func testURL(t *testing.T, r *http.Request, want string) {
-	if c.Account != "AZURE_DEVOPS_ACCOUNT" {
-		t.Errorf("Client.Account = %s; expected %s", c.Account, "AZURE_DEVOPS_ACCOUNT")
+	devOpsClient := azuredevops.NewDevOpsClient("AZURE_DEVOPS_ACCOUNT", "AZURE_DEVOPS_TOKEN")
+	
```
Can you remove the trailing white space please?

benmatselby commented on azuredevops/boards.go:
```diff
@@ -8,11 +8,11 @@ import (
-// ListBoardsResponse describes the boards response
-type ListBoardsResponse struct {
+// BoardsListResponse describes the boards response
+type BoardsListResponse struct {
```
Just out of interest, is this a rename because it's in the Boards service? Curious.

pbelousov replied:
I did this in order to be more consistent with other ListResponse types - BuildDefinitionsListResponse, BuildsListResponse, DeliveryPlansListResponse, etc.

pbelousov commented:
Sorry for the long delay and for the broken tests. Had forgotten that these changes will be applied to this PR. Fixed mentioned issues and added a test for the new method.

mcdafydd commented:
Greetings!  First, thank you for the awesome start into Go.  This repository led to my first project in the language, adding webhook support in my fork.  I agree with this change allowing the project to be separated.

I have been porting a lot of the style and approaches taken by the go-github repo.  My NewClient() is a near clone of that where it's just an http client, and then all of the API functions now include the org and project in the function parameters:

https://github.com/mcdafydd/go-azuredevops/blob/master/examples/createandmerge/main.go#L74
https://github.com/mcdafydd/go-azuredevops/blob/master/azuredevops/git.go#L441

I like the idea of having a near-complete go-github clone for Azure Devops.  I think there would be value in it for people and companies invested in the Azure/Azure Devops ecosystem.

Thanks much

mcdafydd commented:
I think from the standpoint of a user consuming the library, having a single interface to NewClient is probably easier than trying to make them figure out when they need one versus the other.  What do you think?

Best regards
